### PR TITLE
Fix issues 4697 and 2461 (`@JsonUnwrapped` caching)

### PR DESCRIPTION
**Pull request by SandeepGaur2016**
I recently raised a bug ([4697](https://github.com/FasterXML/jackson-databind/issues/4697)) and have now had some time to investigate the issue. Although this is my first contribution, I noticed that when we obtain a serializer from the case, the name transformation is reset in the _findAndAddDynamic method of UnwrappingBeanPropertyWriter. However, this method does not account for inner or nested properties. I have implemented a fix to reinitialize the serializer in such cases.

It has also fixed looks same issue from raised earlier (#2461)

Thanks,
Sandeep Gaur

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java [hunk 1]
```diff
@@ -24,6 +24,8 @@
 import com.fasterxml.jackson.databind.util.ClassUtil;
 import com.fasterxml.jackson.databind.util.NameTransformer;
 
+import static com.fasterxml.jackson.databind.ser.impl.PropertySerializerMap.emptyForProperties;
```
Comment (JooHyukKim) [inline src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:27]: Let's revert this change and below back to how it was.
This is not related to the PR.

```java
_dynamicSerializers = (ser == null) ? emptyForProperties() : null;
```

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java [hunk 2]
```diff
@@ -459,7 +470,7 @@ Object readResolve() {
             _field = null;
         }
         if (_serializer == null) {
-            _dynamicSerializers = PropertySerializerMap.emptyForProperties();
+            _dynamicSerializers = emptyForProperties();
```
Comment (JooHyukKim) [inline src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:473]: Here also, unncessary change (at least for this PR).

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java [hunk 3]
```diff
@@ -325,7 +326,12 @@ protected BeanPropertyWriter(BeanPropertyWriter base, PropertyName name) {
                     base._internalSettings);
         }
         _cfgSerializationType = base._cfgSerializationType;
-        _dynamicSerializers = base._dynamicSerializers;
+        if(base instanceof UnwrappingBeanPropertyWriter) {
+            _dynamicSerializers =
+                    emptyForProperties();
+        } else {
+            _dynamicSerializers = base._dynamicSerializers;
+        }
```
Comment (JooHyukKim) [inline src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:334]: Keep the format consistent.

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java [hunk 4]
```diff
@@ -350,8 +356,13 @@ protected BeanPropertyWriter(BeanPropertyWriter base, SerializedString name) {
...
         _cfgSerializationType = base._cfgSerializationType;
-        _dynamicSerializers = base._dynamicSerializers;
-        _suppressNulls = base._suppressNulls;
+        if(base instanceof UnwrappingBeanPropertyWriter) {
+            _dynamicSerializers =
+                    emptyForProperties();
+        } else {
+            _dynamicSerializers = base._dynamicSerializers;
+        }
+		_suppressNulls = base._suppressNulls;
```
Comment (JooHyukKim) [inline src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:365]: Formatting needs to be done.

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java [hunk 5]
```diff
@@ -325,7 +326,12 @@ protected BeanPropertyWriter(BeanPropertyWriter base, PropertyName name) {
                     base._internalSettings);
         }
         _cfgSerializationType = base._cfgSerializationType;
-        _dynamicSerializers = base._dynamicSerializers;
+        if(base instanceof UnwrappingBeanPropertyWriter) {
```
Comment (JooHyukKim) [inline src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:329]: Dependency from super class to subclasses doesn't seem to be the right direction.

**Review by JooHyukKim — COMMENTED** [review]
Added a few notes, thanks!

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java (hunk 1 above)
Comment (SandeepGaur2016) [reply src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:27]: done, though it was a clean way to import statically

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java (hunk 2 above)
Comment (SandeepGaur2016) [reply src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java]: done, though it was a clean way to import statically

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java (hunk 3 above)
Comment (SandeepGaur2016) [reply src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java]: Agreed, I will set the formatter.

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java (hunk 4 above)
Comment (SandeepGaur2016) [reply src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java]: done

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java (hunk 5 above)
Comment (SandeepGaur2016) [reply src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java]: Yes, did the same thing in a better way now

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java [hunk 6]
```diff
@@ -350,8 +355,12 @@ protected BeanPropertyWriter(BeanPropertyWriter base, SerializedString name) {
...
         }
         _cfgSerializationType = base._cfgSerializationType;
-        _dynamicSerializers = base._dynamicSerializers;
-        _suppressNulls = base._suppressNulls;
+        if (base.needToResetSerialization()) {
+            _dynamicSerializers = PropertySerializerMap.emptyForProperties();
+        } else {
+            _dynamicSerializers = base._dynamicSerializers;
+        }
+		_suppressNulls = base._suppressNulls;
```
Comment (JooHyukKim) [inline src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:363]: Where are these constructors called? Does this mean serializer information from 'base' should be lost?

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java (hunk 2 above)
Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java]: I prefer avoiding static imports for method calls fwtw. Maybe that should be added in style guide (or maybe it is mentioned and CONTRIBUTING.md should refer). I know IDEs can show where method comes from but for me (and I realize this is matter of preference) little bit of redundancy can be helpful.

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java [hunk 7]
```diff
@@ -325,7 +326,11 @@ protected BeanPropertyWriter(BeanPropertyWriter base, PropertyName name) {
                     base._internalSettings);
         }
         _cfgSerializationType = base._cfgSerializationType;
-        _dynamicSerializers = base._dynamicSerializers;
+        if (base.needToResetSerialization()) {
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:329]: I am not sure check is needed, should probably simply reset always; I'll add a general note to PR.

File: src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java (hunk 6 above)
Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java:363]: `_dynamicSerializers` is purely performance optimization, caching to avoid unnecessary full introspection -- should never be needed for correctness. Hence if necessary we can just use empty instance for copy constructors.

File: src/main/java/com/fasterxml/jackson/databind/ser/PropertyWriter.java [hunk 8]
```diff
@@ -100,6 +100,13 @@ public <A extends Annotation> A findAnnotation(Class<A> acls) {
     /**********************************************************
      */
 
+    /**
+     * A way to reset serialization to correctly populate serializer
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/ser/PropertyWriter.java:104]: This unfortunately doesn't really describe anything....

**Comment by cowtowncoder** [comment]
Ok: first of all, yes, it does sound like (and makes sense, too) that the issue is incorrect caching of resolved serializers: `_dynamicSerializers` is strictly for performance optimization. As such, maybe we should just be more aggressively clearing it on copy constructors: I'm fine not requiring checker method for trying to figure out.

Alternatively there is an existing method `BeanPropertyWriter.isUnwrapping()` which could (and for practical reasons I'd argue) and should be called instead of adding a new method.

But it seems to me simplest to just re-initialize -- empty instances are shared stateless singletons so there's no real cost for constructing. And in most cases this makes no difference anyway.

**Comment by SandeepGaur2016** [comment]
Hi, I have removed the condition of UnwrappingBeanPropertyWrite, I simply set the _dynamicSerializers  as empty list.

**Comment by cowtowncoder** [comment]
Ok, happy to merge this fix. Just one process thing first: since this is (I think) your first contribution, we'd need CLA from:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

(this only needs to be done once before the first PR gets merged -- one is good for all future PRs).

It is usually easiest to print, fill & sign, scan/photo, email to `cla` at fasterxml dot com.
Once I get it, I'll get this merged ASAP.

Thank you again @SandeepGaur2016 for the contribution!

**Comment by cowtowncoder** [comment]
CLA received, can proceed!

**Review by cowtowncoder — APPROVED** [review]
